src/coreclr/src/jit/lclvars.cpp

```diff
@@ -3680,18 +3680,20 @@ bool LclVarDsc::CanBeReplacedWithItsField(Compiler* comp) const
     }
 
 #if defined(FEATURE_SIMD)
-    assert(!comp->isOpaqueSIMDLclVar(this)); // opaque SIMD can't be promoted.
     // If we return `struct A { SIMD16 a; }` we split the struct into several fields.
     // In order to do that we have to have its field `a` in memory. Right now lowering cannot
     // handle RETURN struct(multiple registers)->SIMD16(one register), but it can be improved.
     LclVarDsc* fieldDsc = comp->lvaGetDesc(lvFieldLclStart);
     if (fieldDsc->TypeGet() == TYP_SIMD12 || fieldDsc->TypeGet() == TYP_SIMD16)
     {
 #if defined(TARGET_ARM64)
-        return false;
+        if (!comp->isOpaqueSIMDLclVar(fieldDsc))
```

Comment:
I think that this is backward. If it is opaque, that means that it will be passed in a single register on Arm64 (and on x64/ux after #9578 is fixed). For those cases, we can use a single field, e.g. for `struct A { Vector128<T> a; }` However, for `struct A {Vector2 a; }` we can't as it will be passed/returned in multiple registers.

Reply:
I was not sure enough in this change so I stepped back and played with it for a while.

As I found out we need this block for:
-Linux x64 WrappedVector3/4;
-arm64 needs it for Vector2/3/4/T and Vector128.

Linux x64 WrappedVector2 does not need it because we return `SIMD8` as `double`; `Vector<T>` does not need this block because we pass it byref.

Without this block for listed cases when we were receiving IR like:
```
               [000016] ------------              *  RETURN    struct (xmm0, xmm1)
               [000015] -------N----              \--*  LCL_VAR   struct<Vector3Wrapper, 12>(P) V00 loc0
                                                  \--*    simd12 V00.f1 (offs=0x00) -> V03 tmp1											  
```

we were changing it to:
```
               [000016] -----+------              *  RETURN    struct (xmm0, xmm1)
               [000015] -----+-N----              \--*  LCL_VAR   simd12<System.Numerics.Vector3> V03 tmp1
```

and we were failing in `Lowering::ContainCheckRet` or in codegen because `RETURN` and `LCL_VAR` had different number of registers (`LCL_VAR` - 1, `RETURN` > 1).

I was trying to make a special condition for `isOpaqueSIMDLclVar` but it did not work well, for example:
`struct A { Vector<short> f; }` returns `isOpaqueSIMDLclVar` on arm64, but we return it as 2 registers x0, x1. A similar scenario with `Vector128`, it is also 
returned as x0,x1, so even for `isOpaqueSIMDLclVar` we were not returning the wrapping struct as one Vector register. 

So my current solution is to always block this field promotion for SIMD vars, it won't be a regression in comparasing with `oldRetyping` and open an issue to support it better later.

Note: I have checked that some of these new tests were failing with old retyping before, meaning that we have not seen these patterns often.

Reply:
On Arm64, vectors are passed in a single register, and it sounds like we need to fix the altjit to do the right thing.

> my current solution is to always block this field promotion for SIMD vars, it won't be a regression in comparasing with oldRetyping and open an issue to support it better later.

That seems reasonable, though I hope that we can address that issue in the near term.